File: src/sp/it/pl/gui/itemnode/textfield/EffectItemNode.java
package sp.it.pl.gui.itemnode.textfield;

import java.lang.reflect.InvocationTargetException;
import java.util.List;
import javafx.scene.control.Tooltip;
import javafx.scene.effect.Blend;
import javafx.scene.effect.Bloom;
import javafx.scene.effect.BoxBlur;
import javafx.scene.effect.ColorAdjust;
import javafx.scene.effect.ColorInput;
import javafx.scene.effect.DisplacementMap;
import javafx.scene.effect.DropShadow;
import javafx.scene.effect.Effect;
import javafx.scene.effect.GaussianBlur;
import javafx.scene.effect.Glow;
import javafx.scene.effect.ImageInput;
import javafx.scene.effect.InnerShadow;
import javafx.scene.effect.Lighting;
import javafx.scene.effect.MotionBlur;
import javafx.scene.effect.PerspectiveTransform;
import javafx.scene.effect.Reflection;
import javafx.scene.effect.SepiaTone;
import javafx.scene.effect.Shadow;
import javafx.scene.input.MouseEvent;
import sp.it.pl.gui.objects.icon.Icon;
import sp.it.pl.gui.objects.picker.Picker;
import sp.it.pl.gui.objects.popover.PopOver;
import sp.it.pl.util.type.ClassName;
import static javafx.geometry.Pos.CENTER;
import static sp.it.pl.main.AppUtil.APP;
import static sp.it.pl.main.AppBuildersKt.appTooltip;
import static sp.it.pl.util.conf.Configurable.configsFromFxPropertiesOf;
import static sp.it.pl.util.dev.Util.log;
import static sp.it.pl.util.functional.Util.list;
import static sp.it.pl.util.functional.Util.stream;
import static sp.it.pl.util.graphics.Util.layHorizontally;

public class EffectItemNode extends TextFieldItemNode<Effect> {

	public static class EffectType {
		/** Effect class. Can be null. Null represents no effect. */
		public final Class<? extends Effect> type;

		public EffectType(Class<? extends Effect> type) {
			this.type = type;
		}

		String name() {
			return type==null ? "None" : ClassName.of(type);
		}

		Effect instantiate() {
			try {
				return type==null ? null : type.getConstructor().newInstance();
			} catch (InstantiationException|IllegalAccessException|NoSuchMethodException|InvocationTargetException x) {
				log(EffectItemNode.class).error("Config could not instantiate effect", x);
				return null;
			}
		}
	}

	private static final Tooltip typeTooltip = appTooltip("Choose type of effect");
	private static final Tooltip propTooltip = appTooltip("Configure effect");
	public static final List<EffectType> EFFECT_TYPES = list(
			new EffectType(Blend.class),
			new EffectType(Bloom.class),
			new EffectType(BoxBlur.class),
			new EffectType(ColorAdjust.class),
			new EffectType(ColorInput.class),
			new EffectType(DisplacementMap.class),
			new EffectType(DropShadow.class),
			new EffectType(GaussianBlur.class),
			new EffectType(Glow.class),
			new EffectType(ImageInput.class),
			new EffectType(InnerShadow.class),
			new EffectType(Lighting.class),
			new EffectType(MotionBlur.class),
			new EffectType(PerspectiveTransform.class),
			new EffectType(Reflection.class),
			new EffectType(SepiaTone.class),
			new EffectType(Shadow.class),
			new EffectType(null)
	);

	private final Icon typeB = new Icon();
	private final Icon propB = new Icon();
	private final Class<? extends Effect> limitedToType;

	/** Creates effect editor which can edit an effect or create effect of any type. */
	public EffectItemNode() {
		this(null);
	}

	/** Creates effect editor which can edit an effect or create effect of specified type. */
	public EffectItemNode(Class<? extends Effect> effectTypeLimit) {
		super(et -> ClassName.of(et.getClass()));
		limitedToType = effectTypeLimit==Effect.class ? null : effectTypeLimit;
		setEditable(false);

		typeB.styleclass("effect-config-field-type-button");
		typeB.tooltip(typeTooltip);
		typeB.onClick(this::openChooser);
		propB.styleclass("effect-config-field-conf-button");
		propB.tooltip(propTooltip);
		propB.onClick(this::openProperties);

		setRight(layHorizontally(5, CENTER, typeB, propB));

		// initialize value & graphics
		setValue(null);
	}

	@Override
	void onDialogAction() {}

	@Override
	public void setValue(Effect value) {
		propB.setDisable(value==null);
		super.setValue(value);
	}

	private void openChooser(MouseEvent me) {
		Picker<EffectType> w = new Picker<>();
		w.itemSupply = limitedToType==null ? EFFECT_TYPES::stream : () -> stream(new EffectType(limitedToType), new EffectType(null));
		w.textConverter = EffectType::name;
		w.getNode().setPrefSize(300, 500);
		PopOver<?> p = new PopOver<>(w.getNode());
		p.title.set("Effect");
		p.setArrowSize(0);
		p.setAutoFix(true);
		p.setAutoHide(true);
		p.show(propB);
		w.onCancel = p::hide;
		w.onSelect = ec -> {
			setValue(ec.instantiate());
			openProperties(me);
			p.hide();
		};
	}

	private void openProperties(MouseEvent me) {
		if (v!=null) APP.windowManager.showSettings(configsFromFxPropertiesOf(v), me);
		me.consume();
	}
}

File: src/sp/it/pl/gui/itemnode/textfield/EffectItemNode.kt
package sp.it.pl.gui.itemnode.textfield

import javafx.event.EventHandler
import javafx.geometry.Pos.CENTER
import javafx.scene.effect.Blend
import javafx.scene.effect.Bloom
import javafx.scene.effect.BoxBlur
import javafx.scene.effect.ColorAdjust
import javafx.scene.effect.ColorInput
import javafx.scene.effect.DisplacementMap
import javafx.scene.effect.DropShadow
import javafx.scene.effect.Effect
import javafx.scene.effect.GaussianBlur
import javafx.scene.effect.Glow
import javafx.scene.effect.ImageInput
import javafx.scene.effect.InnerShadow
import javafx.scene.effect.Lighting
import javafx.scene.effect.MotionBlur
import javafx.scene.effect.PerspectiveTransform
import javafx.scene.effect.Reflection
import javafx.scene.effect.SepiaTone
import javafx.scene.effect.Shadow
import javafx.scene.input.MouseEvent
import javafx.scene.layout.Region
import mu.KLogging
import sp.it.pl.gui.objects.icon.Icon
import sp.it.pl.gui.objects.picker.Picker
import sp.it.pl.gui.objects.popover.PopOver
import sp.it.pl.main.AppUtil.APP
import sp.it.pl.main.appTooltip
import sp.it.pl.util.conf.Configurable.configsFromFxPropertiesOf
import sp.it.pl.util.functional.Util.stream
import sp.it.pl.util.graphics.Util.layHorizontally
import sp.it.pl.util.type.ClassName
import java.lang.reflect.InvocationTargetException
import java.util.function.Consumer
import java.util.function.Function
import java.util.function.Supplier
import kotlin.reflect.KClass

class EffectItemNode: TextFieldItemNode<Effect> {

    private val typeB: Icon
    private val propB: Icon
    private val limitedToType: Class<out Effect>?

    /** Creates effect editor which can edit an effect or create effect of any specified types or any type if no specified. */
    @JvmOverloads constructor(effectType: Class<out Effect>? = null): super({ ClassName.of(it.javaClass) }) {
        typeB = Icon().apply {
            styleclass("effect-config-field-type-button")
            tooltip(typeTooltip)
            onClick(EventHandler { openChooser(it) })
        }
        propB = Icon().apply {
            styleclass("effect-config-field-conf-button")
            tooltip(propTooltip)
            onClick(EventHandler { openProperties(it) })
        }
        limitedToType = if (effectType==Effect::class.java) null else effectType
        isEditable = false
        right.value = layHorizontally(5.0, CENTER, typeB, propB)

        value = null    // initialize value & graphics
    }

    override fun onDialogAction() {}

    override fun setValue(value: Effect?) {
        propB.isDisable = value==null
        super.setValue(value)
    }

    private fun openChooser(me: MouseEvent) {
        val p = PopOver<Region>().apply {
            title.set("Effect")
            arrowSize.value = 0.0
            isAutoFix = true
            isAutoHide = true
            contentNode.value = Picker<EffectType>().apply {
                itemSupply = limitedToType
                        ?.let { Supplier { stream(EffectType(limitedToType.kotlin), EffectType(null)) } }
                        ?: Supplier { EFFECT_TYPES.stream() }
                textConverter = Function { it.name() }
                node.setPrefSize(300.0, 500.0)
                onCancel = Runnable { hide() }
                onSelect = Consumer {
                    value = it.instantiate()
                    openProperties(me)
                    hide()
                }
            }.node
        }
        p.showInCenterOf(propB)
    }

    private fun openProperties(me: MouseEvent) {
        if (v!=null) APP.windowManager.showSettings(configsFromFxPropertiesOf(v), me)
        me.consume()
    }

    companion object: KLogging() {

        private val typeTooltip = appTooltip("Choose type of effect")
        private val propTooltip = appTooltip("Configure effect")
        @JvmField val EFFECT_TYPES = listOf(
                EffectType(Blend::class),
                EffectType(Bloom::class),
                EffectType(BoxBlur::class),
                EffectType(ColorAdjust::class),
                EffectType(ColorInput::class),
                EffectType(DisplacementMap::class),
                EffectType(DropShadow::class),
                EffectType(GaussianBlur::class),
                EffectType(Glow::class),
                EffectType(ImageInput::class),
                EffectType(InnerShadow::class),
                EffectType(Lighting::class),
                EffectType(MotionBlur::class),
                EffectType(PerspectiveTransform::class),
                EffectType(Reflection::class),
                EffectType(SepiaTone::class),
                EffectType(Shadow::class),
                EffectType(null)
        )
    }

    class EffectType {

        /** Effect type. Null represents no effect.  */
        val type: Class<out Effect>?

        constructor(type: KClass<out Effect>?) {
            this.type = type?.java
        }

        internal fun name(): String = if (type==null) "None" else ClassName.of(type)

        internal fun instantiate(): Effect? = try {
            type?.getConstructor()?.newInstance()
        } catch (e: InstantiationException) {
            logger.error(e) { "Config could not instantiate effect" }
            null
        } catch (e: IllegalAccessException) {
            logger.error(e) { "Config could not instantiate effect" }
            null
        } catch (e: NoSuchMethodException) {
            logger.error(e) { "Config could not instantiate effect" }
            null
        } catch (e: InvocationTargetException) {
            logger.error(e) { "Config could not instantiate effect" }
            null
        }
    }

}